File: fsharp/sieve/Sieve.fs
module Sieve

let primes limit = 
    let max = float limit |> sqrt |> floor |> int
    let sieve excluded current = 
        if List.contains current excluded 
            then excluded 
            else excluded @ [2*current..current..limit]
    let excluded = [2..max] |> Seq.fold sieve []
    Seq.except excluded [2..limit] |> Seq.toList

File: go/sieve/sieve.go
package sieve

// Sieve returns all primes up to the limit
func Sieve(limit int) []int {
	nonprimes := map[int]bool{}
	result := []int{}
	for i := 2; i <= limit; i++ {
		if nonprimes[i] {
			continue
		}
		result = append(result, i)
		for j := i * 2; j <= limit; j += i {
			nonprimes[j] = true
		}
	}
	return result
}
